pkg/agent/monitoraddnodes.go

```diff
@@ -0,0 +1,453 @@
+		if mon.in(HostInstallationStarted) &&
+			mon.nodeCommandOutputContains("ps -ef | grep coreos-installer", "coreos-installer") {
+			mon.setState(CoreosInstallerWritingToDisk)
+		}
+	},
+	int(HostPreparingForFirstReboot): func(mon *addNodeMonitor) {
+		if mon.in(CoreosInstallerWritingToDisk) &&
+			!mon.cluster.CanSSHToNodeZero() {
+			mon.setState(HostPreparingForFirstReboot)
+		}
+	},
+	int(AfterFirstReboot): func(mon *addNodeMonitor) {
+		if mon.before(AfterFirstReboot) &&
+			mon.nodeCommandOutputContains("ps -ef", "ostree") {
+			mon.setState(AfterFirstReboot)
+		}
+	},
+	int(AfterSecondReboot): func(mon *addNodeMonitor) {
+		if mon.before(AfterSecondRebootKubeletIsUp) &&
+			mon.nodeCommandOutputContains("ps -ef", "coredns") {
+			mon.setState(AfterSecondReboot)
+		}
+	},
+	int(AfterSecondRebootKubeletIsUp): func(mon *addNodeMonitor) {
+		if mon.onOrAfter(AfterSecondReboot) &&
+			mon.before(FirstCSRPendingApproval) &&
+			mon.nodeCommandOutputContains("ps -ef", "kubelet") {
+			mon.setState(AfterSecondRebootKubeletIsUp)
+		}
+	},
+	int(FirstCSRPendingApproval): func(mon *addNodeMonitor) {
+		if mon.onOrAfter(AfterSecondReboot) &&
+			mon.onOrBefore(FirstCSRPendingApproval) &&
+			mon.nodeCommandOutputContains("sudo KUBECONFIG=/etc/kubernetes/kubeconfig oc get csr | grep Pending", "node-bootstrapper") {
+			mon.setState(FirstCSRPendingApproval)
+			mon.logCSRsPendingApproval()
+		}
+	},
+	int(SecondCSRPendingApproval): func(mon *addNodeMonitor) {
+		if mon.onOrAfter(AfterSecondReboot) &&
+			mon.onOrBefore(SecondCSRPendingApproval) &&
+			mon.nodeCommandOutputContains("sudo KUBECONFIG=/etc/kubernetes/kubeconfig oc get csr | grep Pending", "system:node") {
+			mon.setState(SecondCSRPendingApproval)
+			mon.logCSRsPendingApproval()
+		}
+	},
+	int(NodeJoinedClusterStatusNotReady): func(mon *addNodeMonitor) {
+		if mon.in(AfterSecondRebootKubeletIsUp) ||
+			mon.in(SecondCSRPendingApproval) {
+			hasJoined, isReady, err := mon.nodeHasJoinedClusterAndIsReady()
+			if err != nil {
+				logrus.Debugf("HasNodeJoinedClusterAndIsReady returned err: %v", err)
+			}
+			if hasJoined && !isReady {
+				mon.setState(NodeJoinedClusterStatusNotReady)
+			}
+		}
+	},
+	int(Finish): func(mon *addNodeMonitor) {
+		if mon.in(AfterSecondRebootKubeletIsUp) ||
+			mon.in(SecondCSRPendingApproval) ||
+			mon.in(NodeJoinedClusterStatusNotReady) {
+			hasJoined, isReady, err := mon.nodeHasJoinedClusterAndIsReady()
+			if err != nil {
+				logrus.Debugf("HasNodeJoinedClusterAndIsReady returned err: %v", err)
+			}
+			if hasJoined && isReady {
+				mon.setState(Finish)
+			}
+		}
+	},
+}
+
+// MonitorAddNodes waits for the a node to be added to the cluster
+// and reports its status until it becomes Ready.
+func MonitorAddNodes(cluster *Cluster, nodeIPAddress string) error {
+	timeout := 90 * time.Minute
+	waitContext, cancel := context.WithTimeout(cluster.Ctx, timeout)
+	defer cancel()
+
+	mon := newAddNodeMonitor(nodeIPAddress, cluster)
+
+	wait.Until(func() {
```

Comment:
I think here we should really find a way to reuse (most of the all) code already present in [IsBootstrapComplete()](https://github.com/openshift/installer/blob/30ae4204715580d9f4bcc8d7ba58e8131511533e/pkg/agent/cluster.go#L122). Even though there's not bootstrap cluster in the addNodes workflow (because the cluster already exists), all the logic in the [agentRestAPILive](https://github.com/openshift/installer/blob/30ae4204715580d9f4bcc8d7ba58e8131511533e/pkg/agent/cluster.go#L178C5-L178C21) block remains fully valid (until the point where the host installation kicks off).
In particular is fundamental to print out in the console any potential pre-flight validation failures from Assisted-service, which will help the user to troubleshoot errors that could prevent the host to be installed as a new node (and the existing code should already take care of that).
Then, after the host installation has been successfully started, it would be fine to have some custom logic - specific to the monitor add command - to understand at which point the process is, up to waiting for the two CSRs.

Reply:
That would be useful information to have. I've added the assisted-service validation and events to monitor-add-nodes.